### ISSUE TITLE
Scope rules confuse me

### ISSUE THREAD

**Issue by peter-ahe-google** [dart-lang/sdk#28169]
The specification says:

>It is a compile-time error if there is more than one entity with the same name declared in the same scope.

...

>Dart is lexically scoped.    Scopes may nest.

...

>The formal parameter list of a function introduces a new scope known as the function's  *formal parameter scope.*

...

>However, the scope of a function's signature is the function's enclosing scope, not the formal parameter scope.

Lexically scope implies to me that a scope a scope has an extend that corresponds strictly to a range of source code, starting a certain offset in the source text, and ending at another offset in the source text.

For example:

```
{ var x; }
0123456789
```

Here the scope of `x` is from character offset 0 to offset 9 (both ends inclusive, as are all examples below). This makes sense as the declaration of `x` is contained within that range.

Now consider this example:

```
f([f=f]) => f;
01234567890123
          1
```

Here the scope of the formal parameter `f` is 9 to 13. This confuses me as the declaration of `f` isn't contained within that range. However, I can't add an additional parameter named `f`, so it's perhaps more correct to say that parameter's scope is 1 to 13, except the range 4-5. That's a weird scope, because it has a hole where the surrounding scope is restored.

Also, consider this example:

```
var x = []; for (var x in x) {}
0123456789012345678901234567890
          1         2         3
```

Here the scope of the variable `x` declared in the for-in loop is 29-30. Again, the declaration is outside the character range of the scope. I also find that confusing.

Finally, consider this example:

```
// Will cause a compile-time error.
var x = []; for (var x = x.length; x < 10; x++) {}
01234567890123456789012345678901234567890123456789
          1         2         3         4
```

Here the scope of `x` in the for-loop is 12 to 49, with some additional wrinkles (as a new copy of `x` is introduced on each iteration).

I suggest that we change the scope rules for function signatures and for-in loops so that the above examples aren't legal any more, or that we at least issue a warning.

Alternatively, we can accept that scopes around declarations are weird, and also allow this:

```
{ var x; { var x = x; } }
0123456789012345678901234
          1         2
```

We would then say that the scope of the second `x` is 20 to 22, not 9 to 22.

**Comment by lrhn** [comment]
On Wed, Dec 21, 2016 at 4:59 PM, Peter von der Ahé <notifications@github.com
> wrote:

> The specification says:
>
> It is a compile-time error if there is more than one entity with the same
> name declared in the same scope.
>
> ...
>
> Dart is lexically scoped. Scopes may nest.
>
> ...
>
> The formal parameter list of a function introduces a new scope known as
> the function's *formal parameter scope.*
>
> ...
>
> However, the scope of a function's signature is the function's enclosing
> scope, not the formal parameter scope.
>
> Lexically scope implies to me that a scope a scope has an extend that
> corresponds strictly to a range of source code, starting a certain offset
> in the source text, and ending at another offset in the source text.
>

That's a very physical definition. I don't actually think of source
characters, only abstract syntax trees.

> For example:
>
> { var x; }
> 0123456789
>
> Here the scope of x is from character offset 0 to offset 9 (both ends
> inclusive, as are all examples below). This makes sense as the declaration
> of x is contained within that range.
>
This is an embedded "declaration statement". It introduces a variable into
a scope, in this case the scope of the block that it's declared in.
Whether it covers the `{` and `}` *characters* isn't really meaningful to
me.


> Now consider this example:
>
> f([f=f]) => f;
> 01234567890123
>           1
>
> Here the scope of the formal parameter f is 9 to 13.
>
The formal parameter introduces a variable into the parameter scope. That
scope is the current scope of the body body of the function, which is
(equivalent to) `{ return f; }`.

The function declaration is the syntactic construct that contains both the
formal parameter declaration and the body, and it introduces the parameter
scope and makes it apply to the body.


> This confuses me as the declaration of f isn't contained within that
> range.
>
True. Not all declarations are embedded in the same syntax as the scope
they apply to.

> However, I can't add an additional parameter named f, so it's perhaps
> more correct to say that parameter's scope is 1 to 13, except the range
> 4-5. That's a weird scope, because it has a hole where the surrounding
> scope is restored.
>
That is a weird scope, and I don't think thinking about it in that way is
helpful.
You are not allowed two parameters named `f` because that would introduce
two `f` entries in the parameter scope, and we don't allow that. It's no
more or less complex than that.
Looking at the syntax at the character level isn't giving you anything
useful here.


> Also, consider this example:
>
> var x = []; for (var x in x) {}
> 0123456789012345678901234567890
>           1         2         3
>
> Here the scope of the variable x declared in the for-in loop is 29-30.
> Again, the declaration is outside the character range of the scope. I also
> find that confusing.
>
Yes, same thing as the function parameter. The `for` statement *contains* a
declaration, and it introduces a scope based on that declaration, and it
applies that scope to a different sub-syntax than the declaration itself.
Which is a good thing, IMO.


> Finally, consider this example:
>
> // Will cause a compile-time error.
> var x = []; for (var x = x.length; x < 10; x++) {}
> 01234567890123456789012345678901234567890123456789
>           1         2         3         4
>
> Here the scope of x in the for-loop is 12 to 49, with some additional
> wrinkles (as a new copy of x is introduced on each iteration).
>
I don't think the current spec is very good at distinguishing static scopes
and "runtime scopes". There is one static scope here, but one runtime scope
(that is: one variable named 'x') per iteration.

The extra variables are introduces through renaming, but we could also have
introduced a new scope for the update expression, and moved values between
the scopes.

We are treating the declaration in the initializer part as other "statement
declarations", which means that it is in scope for its own initializer
expression, but you are not allowed to refer to it.
I never liked that design, to be honest.

> I suggest that we change the scope rules for function signatures and
> for-in loops so that the above examples aren't legal any more, or that we
> at least issue a warning.
>
Those two are actually the cases that I like. I'd rather change for(;;) and
declaration statements to not shadow their own initializer expressions. I
find that to be surprising and annoying, because you can't actually *use*
the variable even if it is "in scope".


> Alternatively, we can accept that scopes around declarations are weird,
> and also allow this:
>
> { var x; { var x = x; } }
> 0123456789012345678901234
>           1         2
>
> We would then say that the scope of the second x is 20 to 22, not 9 to 22.
>
I would like that. And it wouldn't even be a breaking change :)

/L
-- 
Lasse R.H. Nielsen - lrn@google.com
'Faith without judgement merely degrades the spirit divine'
Google Denmark ApS - Frederiksborggade 20B, 1 sal - 1360 København K
- Denmark - CVR nr. 28 86 69 84

**Comment by peter-ahe-google** [comment]
> That's a very physical definition. I don't actually think of source
characters, only abstract syntax trees.

Good point. I think trees actually better capture what I was trying to convey.

Please allow me to think out aloud, to formulate some rules about scopes as trees:

1. A scope covers a tree with a single root.

2. Scopes can be nested, but each root has only one scope.

3. A scope can be declarative, in which case a declaration introduces a name in the current scope.

4. A scope can be imperative (or temporal), in which case a declaration introduces a new name in a scope that follows the declaration until the end of the current scope.

5. It is a compile-time error if a name is introduced more than once in a scope.

Currently, some of our scopes are declarative, and some are imperative.

I would say that our current formal parameter scope is imperative, and all parameters must be unique because they all introduce a name in the scope that starts after the formal parameter list.

I can't fit the scope of constructor initializers in this terminology, and I'll just ignore them for now.

Using this terminology, I would restate my first suggestion from above as:

Scopes are declarative.

And my second suggestion as:

Scopes in function, constructor, and factory declarations are imperative, all other scopes are declarative.

As Lasse notes, this wouldn't be a breaking change. In addition, this is what dart2js actually implements.

It has been argued in the past that the latter is faster as it requires only a single pass through code, and the former requires two passes. I'd like to dispel this, as one can simply remember each name *used* (with location) in a scope and report an error if a conflicting name is introduced. I think this is important to note as one can then consider each alternative without worrying about implementation details.

**Comment by lrhn** [comment]
>
> That's a very physical definition. I don't actually think of source
> characters, only abstract syntax trees.
>
> Good point. I think trees actually better capture what I was trying to
> convey.
>
> Please allow me to think out aloud, to formulate some rules about scopes
> as trees:
>
>    1.
>
>    A scope covers a tree with a single root.
>
> All trees have a single root, so that's just a tree :)
I would not have a problem with letting the same scope cover more than one
tree. For example, the parameter scope of a generative constructor covers
both the initializer list and the body. Arguably, that's just a matter of
grouping the syntax tree, you could put initializer list and body in a
shared sub-tree of the constructor node instead of having them as direct
child nodes.



>
>    1.
>    2.
>
>    Scopes can be nested, but each root has only one scope.
>
> Agree. That's a way of saying that scopes *must* be nested - if you have
two scopes applying to a tree, one must be considered to be inside the
other.


>
>    1.
>    2.
>
>    A scope can be declarative, in which case a declaration introduces a
>    name in the current scope.
>
> That's what most top-level and class-level declarations do, but also what
declaration-statements do inside code blocks.

I think I would prefer to think of it as the declaration being declarative,
not the scope.

The scope of a function body is the parameter scope, and the parameters are
not declarative, but declaration-statements inside the body do add to that
scope.
(It can be argued that the `{}` around the body introduces a new block
scope, but I don't think that artificial distinction should change
anything).


>
>    1.
>    2.
>
>    A scope can be imperative (or temporal), in which case a declaration
>    introduces a new name in a scope that follows the declaration until the end
>    of the current scope.
>
> I think this is generally associated with parameterization. The parameters
of a function apply to the body, the type parameters of a class applies to
the contents (but also to the bounds!). The try-catch-clause variables are
kind-of-like parameters of the catch body (if you see it as a continuation,
and who doesn't :).

Functions are the classical example. The parameters abstract over something
in the body, so they have to be outside of/separate from the body
(otherwise they would shadow themselves, or some other recursive nonsense).
The parameters describe the static signature of the scope, and the runtime
binding satisfies that signature (sort-of like the relation between a class
and an instance).

The distinction is that a "declarative" declaration includes the value, a
parameterizing one (what you call "imperative") doesn't, it only declares
the name and the value will be provided later - from the outside.
Parameterizing declarations typically introduce a new scope because they
sit at the edge between where those parameters are available and where they
are not.

The `for(var x = e1; e2; e3) s;` loop is the odd one out because it
contains a declaration that is not really declarative in behavior, but it
doesn't look like a parameterization either. I think it might be safest to
just think of that loop as syntactic sugar for something like:

  {
    var x = e1;  // Declarative declaration in new scope.
    while (e2) {
      var $tmp = x;
      { // Yet another scope, this one with a binding per iteration.
        var x = $tmp;  // Just a renaming hack to avoid changing variable
names
        $body: s;  // with continue -> break $body;
        $tmp = x;
      }
      x = $tmp;
      e3;
    }
  }

Then it all makes sense, including the scope covering e1.

That doesn't explain `var x = ["a"]; for (var x in x) print(x);`, on the
other hand.
Here the "right hand side" of the "var x in" is not in scope of that
declaration.
That suggests a different syntactic desugaring of `for (var v in e) s` :

   var $it = e.iterator;
   while ($it.moveNext()) {
     var v = $it.current;
     s;
   }

(which is probably exactly what it is rewritten to).

So, in summary, we have two loop constructs that can contain a declaration,
and how that declaration affects the scope is really just an artifact of
what syntactic sugar that loop corresponds to.


>
>    1.
>    2.
>
>    It is a compile-time error if a name is introduced more than once in a
>    scope.
>
> Currently, some of our scopes are declarative, and some are imperative.
>
Again, I'd say that some of our declarations are declarative (affect their
textually surrounding scope) and some are imperative (affects the scope of
a sibling syntactic construct as determined by the parent construct).


> I would say that our current formal parameter scope is imperative, and all
> parameters must be unique because they all introduce a name in the scope
> that starts after the formal parameter list.
>
Yes. The parameter scope is the scope of the body of the function.


> I can't fit the scope of constructor initializers in this terminology, and
> I'll just ignore them for now.
>
The parameter scope of a generative constructor covers both the initializer
list and the body. As I said above, you can either think of that as one
scope applying to two separate sub-trees or you can design the syntax so
it's a shared subtree (which makes a kind of sense when you see the
redirection clause of a redirecting generative constructor as an
alternative for that entire subtree).

The scope of the parameter list itself, including default parameter
expressions, is just the surrounding type parameter scope. That is also the
scope of instance field initializer expressions.


> Using this terminology, I would restate my first suggestion from above as:
>
> Scopes are declarative.
>
I agree that that would be the suggestion. I don't think it's practical -
the parameterizing declarations are handling a job different from the
declarative declarations, and we need both.

> And my second suggestion as:
>
> Scopes in function, constructor, and factory declarations are imperative,
> all other scopes are declarative.
>
Also type parameter scopes and catch clauses.


> As Lasse notes, this wouldn't be a breaking change. In addition, this is
> what dart2js actually implements.
>

I think that is exactly what we have now (modulo how you interpret for
loops).


I was actually saying that I wanted to *change* declaration statements to
introduce a new scope only after the declaration, not covering its own
initializer expression.
That would allow:

   var x = 42;
   {
     var x = x + 2;
     print(x);  // 44
     x++;
     print(x);  // 45
   }
   print(x);  // 42

It would also allow the initializer of the for(;;) loop to refer to a
variable by the same name.

Currently things are defined so that:

  var x = 42;
  {
    print(x);      // Error, cannot refer to "x" yet.
    var x = 37;
  }

I don't find declarations acting backwards (and yet not being available
anyway) to be at all intuitive, but that might just be me :)

> It has been argued in the past that the latter is faster as it requires
> only a single pass through code, and the former requires two passes. I'd
> like to dispel this, as one can simply remember each name *used* (with
> location) in a scope and report an error if a conflicting name is
> introduced. I think this is important to note as one can then consider each
> alternative without worrying about implementation details.
>
If one thing works significantly better for the user, we'll just have to
find an effective implementation.

**Comment by lrhn** [comment]
> Please allow me to think out aloud, to formulate some rules about scopes
> as trees:
>
>    1.    A scope covers a tree with a single root.

All trees have a single root, so that's just a tree :)

I would not have a problem with letting the same scope cover more than one
tree. For example, the parameter scope of a generative constructor covers
both the initializer list and the body. Arguably, that's just a matter of
grouping the syntax tree, you could put initializer list and body in a
shared sub-tree of the constructor node instead of having them as direct
child nodes.

>    2.   Scopes can be nested, but each root has only one scope.
>

Agree. That's a way of saying that scopes *must* be nested - if you have
two scopes applying to a tree, one must be considered to be inside the
other.


>    3. A scope can be declarative, in which case a declaration introduces a
>    name in the current scope.
>

That's what most top-level and class-level declarations do, but also what
declaration-statements do inside code blocks.

I think I would prefer to think of it as the declaration being declarative,
not the scope.

The scope of a function body is the parameter scope, and the parameters are
not declarative, but declaration-statements inside the body do add to that
scope.
(It can be argued that the `{}` around the body introduces a new block
scope, but I don't think that artificial distinction should change
anything).


>    4. A scope can be imperative (or temporal), in which case a declaration
>    introduces a new name in a scope that follows the declaration until the end
>    of the current scope.

I think this is generally associated with parameterization. The parameters
of a function apply to the body, the type parameters of a class applies to
the contents (but also to the bounds!). The try-catch-clause variables are
kind-of-like parameters of the catch body (if you see it as a continuation,
and who doesn't :).

Functions are the classical example. The parameters abstract over something
in the body, so they have to be outside of/separate from the body
(otherwise they would shadow themselves, or some other recursive nonsense).
The parameters describe the static signature of the scope, and the runtime
binding satisfies that signature (sort-of like the relation between a class
and an instance).

The distinction is that a "declarative" declaration includes the value, a
parameterizing one (what you call "imperative") doesn't, it only declares
the name and the value will be provided later - from the outside.
Parameterizing declarations typically introduce a new scope because they
sit at the edge between where those parameters are available and where they
are not.

The `for(var x = e1; e2; e3) s;` loop is the odd one out because it
contains a declaration that is not really declarative in behavior, but it
doesn't look like a parameterization either. I think it might be safest to
just think of that loop as syntactic sugar for something like:
```dart
  {
    var x = e1;  // Declarative declaration in new scope.
    while (e2) {
      var $tmp = x;
      { // Yet another scope, this one with a binding per iteration.
        var x = $tmp;  // Just a renaming hack to avoid changing variable names
        $body: s;  // with continue -> break $body;
        $tmp = x;
      }
      x = $tmp;
      e3;
    }
  }
```
Then it all makes sense, including the scope covering `e1`.

That doesn't explain `var x = ["a"]; for (var x in x) print(x);`, on the
other hand.
Here the "right hand side" of the "var x in" is not in scope of that
declaration.
That suggests a different syntactic desugaring of `for (var v in e) s` :
```dart
   var $it = e.iterator;
   while ($it.moveNext()) {
     var v = $it.current;
     s;
   }
```
(which is probably exactly what it is rewritten to).

So, in summary, we have two loop constructs that can contain a declaration,
and how that declaration affects the scope is really just an artifact of
what syntactic sugar that loop corresponds to.

>    5. It is a compile-time error if a name is introduced more than once in a
>    scope.
>
> Currently, some of our scopes are declarative, and some are imperative.

Again, I'd say that some of our declarations are declarative (affect their
textually surrounding scope) and some are imperative (affects/introduces the scope of
a sibling syntactic construct as determined by the parent construct). 

> I would say that our current formal parameter scope is imperative, and all
> parameters must be unique because they all introduce a name in the scope
> that starts after the formal parameter list.

Yes. The parameter scope is the scope of the body of the function.

> I can't fit the scope of constructor initializers in this terminology, and
> I'll just ignore them for now.

The parameter scope of a generative constructor covers both the initializer
list and the body. As I said above, you can either think of that as one
scope applying to two separate sub-trees or you can design the abstract syntax so
it's a shared subtree (which makes a kind of sense when you see the
redirection clause of a redirecting generative constructor as an
alternative for that entire subtree).

The scope of the parameter list itself, including default parameter
expressions, is just the surrounding type parameter scope. That is also the
scope of instance field initializer expressions.

> Using this terminology, I would restate my first suggestion from above as:
>
> Scopes are declarative.

I agree that that would be the suggestion. I don't think it's practical -
the parameterizing declarations are handling a job different from the
declarative declarations, and we need both.

> And my second suggestion as:
>
> Scopes in function, constructor, and factory declarations are imperative,
> all other scopes are declarative.

Also setters, type parameters and catch clauses.

> As Lasse notes, this wouldn't be a breaking change. In addition, this is
> what dart2js actually implements.
>

I think that is exactly what we have now (modulo how you interpret the `for` loops).

I was actually saying that I wanted to *change* declaration statements to
introduce a new scope only after the declaration, not covering its own
initializer expression.
That would allow:
```dart
   var x = 42;
   {
     var x = x + 2;
     print(x);  // 44
     x++;
     print(x);  // 45
   }
   print(x);  // 42
```
It would also allow the initializer of the `for(;;)` loop to refer to a
variable by the same name.

Currently things are defined so that:
```dart
  var x = 42;
  {
    print(x);      // Error, cannot refer to "x" yet.
    var x = 37;
  }
```
I don't find declarations acting backwards (and yet not being available
anyway) to be at all intuitive, but that might just be me :)

> It has been argued in the past that the latter is faster as it requires
> only a single pass through code, and the former requires two passes. I'd
> like to dispel this, as one can simply remember each name *used* (with
> location) in a scope and report an error if a conflicting name is
> introduced. I think this is important to note as one can then consider each
> alternative without worrying about implementation details.

If one thing works significantly better for the user, we'll just have to
find an effective implementation. Still, it's good to know that finding one isn't even hard :)

**Comment by kmillikin** [comment]
For the record, the desugaring of `for` into `while` is more subtle.

The initializer declares a list of variables, and they can capture ones earlier in the list, and they are mutable.  We need that

```dart
for (var x = 0; y = () => x; ...) {
  x = 1;
  ... y() ...
}
```

works correctly which implies that the initializers have to be evaluated in the scope of the first iteration.

Similarly, the updates can contain subexpressions that close over variables declared in the loop header, so we need to ensure that they are evaluated in the scope of the _next_ (just starting) iteration, not the _current_ (just finishing) one.

What I came up with is that:

```dart
label: for (Type x = init, ...; cond; update, ...) body
```

is translated to:

```dart
{
  bool first = true;
  Type temp, ...;
  label: while (true) {
    Type x, ...;
    if (first) {
      first = false;
      x = init; ...;
    } else {
      x = temp; ...;
      update; ...;
    }
    if (cond) {
      body;
      temp = x; ...;
    } else {
      break;
    }
  }
}
```

It has the advantage that it doesn't require translating the body to handle break and continue.  It has the disadvantage that there is an extra branch per iteration.

**Comment by peter-ahe-google** [comment]
I think we're mostly in agreement. A few minor comments:

> All trees have a single root, so that's just a tree :)


Yes. I wanted to be absolutely sure that there wasn't any confusion on this point.


> I think that is exactly what we have now (modulo how you interpret the for loops).


I don't think that's what we have now. I would describe local variables as being declarative as how specified, and implemented by the analyzer and the VM. Dart2js just hasn't implemented it correctly, so they behave in the imperative way in dart2js.


> If one thing works significantly better for the user, we'll just have to
find an effective implementation.


That's what I like to hear.

**Comment by peter-ahe-google** [comment]
@kmillikin thank you for making that point about for loops, saving me the trouble :-)

**Comment by lrhn** [comment]
Kevin said:
> For the record, the desugaring of `for` into `while` is more subtle.
> The initializer declares a list of variables, and they can capture ones earlier in the list, and they are > mutable.  

That would still work "correctly" with my rewrite, if the variables actually were as I expected. Turns out they aren't :(

> We need that
> 
> ```dart
> for (var x = 0; y = () => x; ...) {
>   x = 1;
>   ... y() ...
> }
> ```
>
> works correctly which implies that the initializers have to be evaluated in the scope of the first iteration.

Ack, it's complicated in a different way from what I remembered. You actually have to evaluate the increment and test expressions on different variables each time. I remembered it as there being one loop variable and one body variable per iteration. Instead there just is one variable per iteration, and the *variable* used in the test and increment parts is the new one each time.

Capturing the variable in the declaration part (or the test or increment parts for that matter) is annoying. I guess that if you know that that doesn't happen, you can make a simpler translation where it doesn't  matter which new variable holds the values when.

Should we consider changing that specification? It seems horribly complicated for what it does. 
The way it's written does reduce the number of variables in play to one per iteration (and not, as I'm pretty sure earlier versions of the spec had, one loop variable plus one per iteration), but the fact that the update and test happens on the next iteration's variable, as it should, also means that the loop structure is weird. It's something that an optimizing compiler would probably prefer to loop-unroll once.

So, in conclusion, `for(;;)` is *weird*.

**Comment by eernstg** [comment]
No milestone now: This is not Dart 2 specific.

**Comment by eernstg** [comment]
I'll close this as stale.
